src/Microsoft.Graph.Core/Requests/Content/BatchRequestContentCollection.cs

```diff
@@ -183,10 +183,69 @@ public BatchRequestContentCollection NewBatchWithFailedRequests(Dictionary<strin
             {
                 if (steps.ContainsKey(response.Key) && !BatchResponseContent.IsSuccessStatusCode(response.Value))
                 {
-                    request.AddBatchRequestStep(steps[response.Key].Request);
+                    var step = steps[response.Key];
+                    var newStep = new BatchRequestStep(
+                        requestId: step.RequestId,
+                        httpRequestMessage: step.Request,
+                        dependsOn: step.DependsOn?.ToList());
+
+                    request.AddBatchRequestStep(newStep);
                 }
             }
             return request;
         }
+
+        /// <summary>
+        /// Creates a new <see cref="BatchRequestContentCollection"/> with all <see cref="BatchRequestStep"/> that failed.
+        /// </summary>
+        /// <param name="responseStatusCodes">A dictionary with response codes, get by executing batchResponseContentCollection.GetResponsesStatusCodesAsync()</param>
+        /// <param name="statusCodesToTreatAsSuccess">Optional additional HTTP status codes to also treat as successful. Success is determined by <see cref="BatchResponseContent.IsSuccessStatusCode(HttpStatusCode)"/> OR membership in this set</param>
+        /// <returns>new <see cref="BatchRequestContentCollection"/> with all failed requests.</returns>
+        public BatchRequestContentCollection NewBatchWithFailedRequests(Dictionary<string, HttpStatusCode> responseStatusCodes, IEnumerable<HttpStatusCode> statusCodesToTreatAsSuccess)
+        {
+            var request = new BatchRequestContentCollection(this.requestAdapter, batchRequestLimit);
+            if (responseStatusCodes == null || responseStatusCodes.Count == 0)
+            {
+                return request;
+            }
+
+            HashSet<HttpStatusCode> successSet = statusCodesToTreatAsSuccess != null
+                ? [.. statusCodesToTreatAsSuccess]
+                : null;
```

Comment:
Why do we have the conversion from IEnumerable to HashSet inside the method? IMO we'd be better served requiring that the HashSet be passed in from the caller.